### PR TITLE
updating metrics queries to work with newer version of kube state met…

### PR DESCRIPTION
**Pull request by laurafitzgerald**
…rics

# Issue link
https://issues.redhat.com/browse/MGDAPI-2634

# What
Changes the following metrics
- `kube_pod_container_resource_limits_memory_bytes` to  `kube_pod_container_resource_requests{resource='memory'}`
- `kube_node_status_capacity_memory_bytes` to `kube_node_status_capacity{resource='memory'}`
- `kube_pod_container_resource_requests_memory_bytes` to `kube_pod_container_resource_requests{resource=memory}`
- `kube_node_status_allocatable_memory_bytes` to `kube_node_status_allocatable{resource='memory'}`

- `kube_node_status_allocatable_cpu_cores` to `kube_node_status_allocatable{resource='cpu'}`
- `kube_pod_container_resource_requests_cpu_cores` to `kube_node_status_allocatable{resource='cpu'}`
- `kube_pod_container_resource_limits_cpu_cores` to `kube_pod_container_resource_limits{resource='cpu'}`

- `kube_node_labels{label_node_role_kubernetes_io_compute='true'}` to `kube_node_role{role='worker'}`

- Removed `by (namespace)` in case of single stat
- Renived `created_by_name` from `kube_pod_info` metric queries

Changes include updates to the following 
Grafana Dashboards
- Cluster Resource
- Resource By Namespace
- Resource By Pod

Prometheus Alert
- AMQOnlineContainerHighMemory 
- ClusterSchedulableMemoryLow
- ClusterSchedulableCPULow
- MultiAZPodDistribution
- ThreeScaleContainerHighMemory
- ThreeScaleContainerHighCPU

Scripts
- capture_resource_metrics.sh

# Verification steps
Verification is required on OpenShift 4.8 and OpenShift pre 4.7.21

4.8 cluster prepared titled `lfitzger-g` in ocm
Pre 4.7.21 cluster is a 4.7.6 cluster titled `lf-476` in ocm

In each of the two clusters provided

Grafana Dashboards
- Cluster Resource
- Resource By Namespace
- Resource By Pod

Each Dashboard should have full metrics expected, e.g. in each memory and cpu table there should appear 5 columns (actual, requested, %actual against requested, limits, %actual against limits) - note deploy pods in 3scale namespace wont return any data, please select another pod.
there should be no single single stat or graphs with no data or N/A or error saying `Only queries that return single series/table is supported`

Alerts
- AMQOnlineContainerHighMemory - only relevant to RHMI which makes me think we should cherry-pick these changes to the 2.9 branch so they could get picked up also in any future RHMI releaes @matskiv wdyt?
- ClusterSchedulableMemoryLow - find the alert in prometheus. click on the query to open it up in the query page and remove `* 85 ` from the end of the queyr to confirm that some data is getting returned.
- ClusterSchedulableCPULow - find the alert in prometheus. click on the query to open it up in the query page and remove `* 85 ` from the end of the queyr to confirm that some data is getting returned.
- MultiAZPodDistribution  - remove `rhoam_version{to_version=""} and ` as this metric is not present running locally and remove `== 1` from the end to show that now data is returned
- ThreeScaleContainerHighMemory - find the alert in prometheus. click on the query to open it up in the query page and remove `> 90` and ensure som date is returned - Note this alert might be firing but this is to be addressed in -> https://issues.redhat.com/browse/MGDAPI-2598
- ThreeScaleContainerHighCPU - find the alert in prometheus. click on the query to open it up in the query page and remove `> 90` and ensure som date is returned

Scripts
Rum tej affected perf test script mentioned in the changes section above in order to  ensure that still works.

Code Base - none of the original queries mentioned above should be found anywhere in the code base.

### CODE REVIEW

**Comment by laurafitzgerald** [comment]
Looks like the test failure might be cluster related so 
/test rhoam-e2e

File: pkg/products/monitoring/prometheusRules.go [hunk 1]
```diff
@@ -294,7 +294,7 @@ func (r *Reconciler) newAlertsReconciler(logger l.Logger, installType string) re
...
 						},
-						Expr:   intstr.FromString(installationName + "_version{to_version=\"\"} and (count by(namespace, created_by_name, label_topology_kubernetes_io_zone) (kube_pod_info{namespace=~'" + nsPrefix + ".*', created_by_kind!=\"<none>\"} == on(node) group_left(label_topology_kubernetes_io_zone) kube_node_labels) == on (namespace, created_by_name)(count by(namespace, created_by_name) (kube_pod_info{namespace=~'" + nsPrefix + ".*', created_by_kind!=\"<none>\"}) > 1) > scalar((count(count by (label_topology_kubernetes_io_zone) (kube_node_labels)) >= bool 2) == 1))"),
+						Expr:   intstr.FromString(installationName + "_version{to_version=\"\"} and (count by(namespace, label_topology_kubernetes_io_zone) (kube_pod_info{namespace=~'" + nsPrefix + ".*', created_by_kind!=\"<none>\"} == on(node) group_left(label_topology_kubernetes_io_zone) kube_node_labels) == on (namespace)(count by(namespace) (kube_pod_info{namespace=~'" + nsPrefix + ".*', created_by_kind!=\"<none>\"}) > 1) > scalar((count(count by (label_topology_kubernetes_io_zone) (kube_node_labels)) >= bool 2) == 1))"),
```
Comment (matskiv) [inline pkg/products/monitoring/prometheusRules.go:297]: @laurafitzgerald `created_by_name` label is not available anymore? 
I think it's likely that query won't work as expected without this.

**Comment by KevFan** [comment]
For 4.8 cluster and pre 4.7.6 cluster:
* Dashboards: `Cluster Resource`, `Resource By Namespace` and `Resource By Pod`
  * Verified each Dashboard have full metrics expected
* Alerts: `ClusterSchedulableMemoryLow` , `ClusterSchedulableCPULow`, `MultiAZPodDistribution`, `ThreeScaleContainerHighMemory` and `ThreeScaleContainerHighCPU`
  * Verified data returned with adjustments to query as per verification steps
* Scripts: `capture_resource_metrics.sh`
  * Verified script completes successfully

I'll let @matskiv put `/lgtm` if he's happy with the changes since he has a question

/test e2e

File: pkg/products/amqonline/prometheusRules.go [hunk 2]
```diff
@@ -146,7 +146,7 @@ func (r *Reconciler) newAlertsReconciler(logger l.Logger, installType string) re
 							"sop_url": resources.SopUrlAlertsAndTroubleshooting,
 							"message": "The {{  $labels.container  }} Container in the {{ $labels.pod  }} Pod has been using {{  $value }}% of available memory for longer than 15 minutes.",
 						},
-						Expr:   intstr.FromString(fmt.Sprintf("sum by(container, pod) (container_memory_usage_bytes{container!='',namespace='%[1]v'}) / sum by(container, pod) (kube_pod_container_resource_limits_memory_bytes{namespace='%[1]v'}) * 100 > 90", r.Config.GetNamespace())),
+						Expr:   intstr.FromString(fmt.Sprintf("sum by(container, pod) (container_memory_usage_bytes{container!='',namespace='%[1]v'}) / sum by(container, pod) (kube_pod_container_resource_requests{namespace='%[1]v'}, resource='memory') * 100 > 90", r.Config.GetNamespace())),
```
Comment (KevFan) [inline pkg/products/amqonline/prometheusRules.go:149]:
```suggestion
						Expr:   intstr.FromString(fmt.Sprintf("sum by(container, pod) (container_memory_usage_bytes{container!='',namespace='%[1]v'}) / sum by(container, pod) (kube_pod_container_resource_requests{namespace='%[1]v', resource='memory'}) * 100 > 90", r.Config.GetNamespace())),
```

Prow e2e looks to fail on installing AMQ online due to invalid prometheus rule here :thinking:

Comment (laurafitzgerald) [reply pkg/products/amqonline/prometheusRules.go:149]: ah. yes. I think i had fixed that change locally. It was failing for me also. Will push it. 👍

File: pkg/products/monitoring/prometheusRules.go (hunk 1 above)
Comment (laurafitzgerald) [reply pkg/products/monitoring/prometheusRules.go:297]: @matskiv yes, i've double checked this against 4.8 cluster and the label is still present and i see the difference in the queries now. Great spot. I will remove this change.

**Comment by laurafitzgerald** [comment]
Final verification can be done on clusters lf-476 and lfitzger
MultiAZPodDistribution - remove rhoam_version{to_version=""} and as this metric is not present running locally and remove == 1 from the end to show that now data is returned

**Comment by KevFan** [comment]
> Final verification can be done on clusters lf-476 and lfitzger
> MultiAZPodDistribution - remove rhoam_version{to_version=""} and as this metric is not present running locally and remove == 1 from the end to show that now data is returned

Verified `MultiAZPodDistribution` returned data with adjustments to query on 4.7.6 and 4.8 cluster :+1: 

/lgtm

**Pull request merged**